lib/vector-core/src/stream/driver.rs

```diff
@@ -0,0 +1,139 @@
+use std::{collections::HashMap, fmt};
+
+use buffers::{Ackable, Acker};
+use futures::{stream::FuturesUnordered, FutureExt, Stream, StreamExt, TryFutureExt};
+use tokio::{pin, select, sync::oneshot};
+use tower::{Service, ServiceExt};
+use tracing::Instrument;
+
+use crate::event::{EventStatus, Finalizable};
+
+/// Drives the interaction between a stream of items and a service which processes them
+/// asynchronously.
+///
+/// `Driver`, as a high-level, facilitates taking items from an arbitrary `Stream` and pushing them
+/// through a `Service`, spawning each call to the service so that work can be run concurrently,
+/// managing waiting for the service to be ready before processing more items, and so on.
+///
+/// Additionally, `Driver` handles two event-specific facilities: finalization and acknowledgement.
+///
+/// This capability is parameterized so any implementation which can define how to interpret the
+/// response for each request, as well as define how many events a request is compromised of, can be
+/// used with `Driver`.
+pub struct Driver<St, Svc> {
+    input: St,
+    service: Svc,
+    acker: Acker,
+}
+
+impl<St, Svc> Driver<St, Svc> {
+    pub fn new(input: St, service: Svc, acker: Acker) -> Self {
+        Self {
+            input,
+            service,
+            acker,
+        }
+    }
+}
+
+impl<St, Svc> Driver<St, Svc>
+where
+    St: Stream,
+    St::Item: Ackable + Finalizable,
+    Svc: Service<St::Item>,
+    Svc::Error: fmt::Debug + 'static,
+    Svc::Future: Send + 'static,
+    Svc::Response: AsRef<EventStatus>,
+{
+    /// Runs the driver until the input stream is exhausted.
+    ///
+    /// All in-flight calls to the provided `service` will also be completed before `run` returns.
+    ///
+    /// # Errors
+    ///
+    /// No errors are currently returned.  The return type is purely to simplify caller code, but may
+    /// return an error for a legitimate reason in the future.
+    pub async fn run(self) -> Result<(), ()> {
+        let mut in_flight = FuturesUnordered::new();
+        let mut pending_acks = HashMap::new();
+        let mut seq_head: u64 = 0;
+        let mut seq_tail: u64 = 0;
+
+        let Self {
+            input,
+            mut service,
+            acker,
+        } = self;
+
+        pin!(input);
+
+        loop {
+            select! {
```

Comment:
Should we bias this `select` towards clearing out `in_flight` before accepting new work? Could help prevent requests from piling up when the input is saturated.

Reply:
Yeah, good idea.  Addressed in 05bc5f31a.